### PR TITLE
Feature/cache middleware

### PR DESCRIPTION
**Pull request by lmammino**
Adds a caching middleware

### CODE REVIEW

**Comment by NPCRUS** [comment]
Correct me if I wrong please. Looks like this is sort of in-memory cache, and in this case we can't really apply such middleware for lambda functions, because containers will have different cache and other problems. Lambda may not save this between executions. It's better to use some sort of redis or memcache in case of using lambda.
So, maybe you should add some sort of notice to this middleware, that it's not for lambda usage

**Comment by peterjcaulfield** [comment]
@NPCRUS I believe @lmammino is intending to leverage container re-use where possible.

**Comment by NPCRUS** [comment]
@peterjcaulfield Thank you very much for this article, make sense!

**Comment by lmammino** [comment]
@NPCRUS, if you check the documentation here and the example you can see that the middleware is broadly configurable. You can provide your own functions to calculate cache ids, save items in the cache and get them from the cache. It should be relatively easy to write those functions to adopt a remote caching backend if you need that.

File: src/middlewares/cache.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+  return ({
+    before: (handler, next) => {
+      options.calculateCacheId(handler.event)
+        .then((cacheKey) => {
+          currentCacheKey = cacheKey
+          return options.getValue(cacheKey)
+        })
+        .then((cachedResponse) => {
+          if (cachedResponse) {
```
Comment (dkatavic) [inline src/middlewares/cache.js]: can we change this with `if (cachedResponse === undefined) {`? . That way we gonna avoid false negatives when cache is returning `0`, `null` or empty string

Comment (lmammino) [reply src/middlewares/cache.js]: Good point :)

**Pull request merged**